homeassistant/components/stream/worker.py

```diff
@@ -160,37 +160,34 @@ def check_flush_part(self, packet: av.Packet) -> None:
             for stream_output in self._outputs_callback().values():
                 stream_output.put(self._segment)
         else:  # These are the ends of the part segments
-            self._segment.parts.append(
-                Part(
-                    duration=float(
-                        (packet.dts - self._part_start_dts) * packet.time_base
-                    ),
-                    has_keyframe=self._part_has_keyframe,
-                    data=self._memory_file.getbuffer()[
-                        self._segment_last_write_pos : byte_position
-                    ].tobytes(),
-                )
-            )
+            self.flush_part(packet, byte_position)
+            # We could put the next two lines in self.flush_part, but they are
+            # already done by reset() in the case of flushing the stub part,
+            # so doing it here avoids doing it twice for the stub part.
             self._segment_last_write_pos = byte_position
             self._part_start_dts = packet.dts
-            self._part_has_keyframe = False
 
-    def flush(self, duration: Fraction, packet: av.Packet) -> None:
-        """Create a segment from the buffered packets and write to output."""
-        self._av_output.close()
+    def flush_part(self, packet: av.Packet, file_pos: int) -> None:
+        """Output a part from the most recent bytes in the memory_file."""
         assert self._segment
-        self._segment.duration = float(duration)
-        # Also flush the part segment (need to close the output above before this)
         self._segment.parts.append(
             Part(
                 duration=float((packet.dts - self._part_start_dts) * packet.time_base),
                 has_keyframe=self._part_has_keyframe,
                 data=self._memory_file.getbuffer()[
-                    self._segment_last_write_pos :
+                    self._segment_last_write_pos : file_pos
```

Comment:
Can the file position always be omitted here? My impression is that this is always flushing the entire buffer every time memory_file is read.

Reply:
Yes, good point. I think it can be omitted.